diff --git a/epan/dissectors/packet-ipsec.c b/epan/dissectors/packet-ipsec.c
index c8699dd8ad..61a8ad4991 100644
--- a/epan/dissectors/packet-ipsec.c
+++ b/epan/dissectors/packet-ipsec.c
@@ -146,6 +146,9 @@ static dissector_table_t ip_dissector_table;
 #define IPSEC_ENCRYPT_AES_GCM_12  10
 #define IPSEC_ENCRYPT_AES_GCM_16  11
 
+/* Encryption algorithm defined in RFC 8750 & RFC 7634*/
+#define IPSEC_ENCRYPT_CHACHA20_POLY1305_IIV 12
+
 /* Authentication algorithms defined in RFC 4305 */
 #define IPSEC_AUTH_NULL 0
 #define IPSEC_AUTH_HMAC_SHA1_96 1
@@ -173,6 +176,7 @@ static dissector_table_t ip_dissector_table;
 #define IPSEC_STRLEN_IPV4 8
 #define IPSEC_SA_IPV4 1
 #define IPSEC_SA_IPV6 2
+#define IPSEC_SA_ANY 3
 #define IPSEC_SA_UNKNOWN -1
 #define IPSEC_SA_WILDCARDS_ANY '*'
 /* the maximum number of bytes (10)(including the terminating nul character(11)) */
@@ -212,6 +216,7 @@ static const value_string esp_encryption_type_vals[] = {
   { IPSEC_ENCRYPT_AES_GCM_8,  "AES-GCM with 8 octet ICV [RFC4106]" },
   { IPSEC_ENCRYPT_AES_GCM_12, "AES-GCM with 12 octet ICV [RFC4106]" },
   { IPSEC_ENCRYPT_AES_GCM_16, "AES-GCM with 16 octet ICV [RFC4106]" },
+  { IPSEC_ENCRYPT_CHACHA20_POLY1305_IIV, "ChaCha20 with Poly1305 IIV [RFC8750]" },
   { 0x00, NULL }
 };
 
@@ -1034,11 +1039,13 @@ filter_address_match(char *addr, char *filter, int typ)
           return false;
       if (!get_full_ipv4_addr(filter_hex, filter))
           return false;
-  } else {
+  } else if(typ == IPSEC_SA_IPV6) {
       if (get_full_ipv6_addr(addr_hex, addr))
           return false;
       if (get_full_ipv6_addr(filter_hex, filter))
           return false;
+  } else if(typ == IPSEC_SA_ANY) {
+      return TRUE;
   }
 
   addr_len = (unsigned)strlen(addr_hex);
@@ -1175,9 +1182,9 @@ get_esp_sa(int protocol_typ, char *src,  char *dst,  unsigned spi,
       record = &uat_esp_sa_records[i++];
     }
 
-    if((protocol_typ == record->protocol)
-       && filter_address_match(src, record->srcIP, protocol_typ)
-       && filter_address_match(dst, record->dstIP, protocol_typ)
+    if((protocol_typ == record->protocol || record->protocol == IPSEC_SA_ANY)
+       && (filter_address_match(src, record->srcIP, protocol_typ) || record->protocol == IPSEC_SA_ANY)
+       && (filter_address_match(dst, record->dstIP, protocol_typ) || record->protocol == IPSEC_SA_ANY)
        && filter_spi_match(spi, record->spi))
     {
       found = true;
@@ -1505,6 +1512,7 @@ dissect_esp(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, void* data _U_)
   int auth_algo_libgcrypt = 0;
   char *esp_icv_expected = NULL; /* as readable hex string, for error messages */
   unsigned char ctr_block[16];
+  unsigned char esp_iiv[12]; /* Implicit IV for decrypting ChaCha20-Poly1805 */
 
 
   uint32_t sequence_number;
@@ -2043,6 +2051,27 @@ dissect_esp(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, void* data _U_)
 
           break;
 
+        case IPSEC_ENCRYPT_CHACHA20_POLY1305_IIV:
+          esp_iv_len = 0; // Implicit IV - First Byte after SEQ is Data
+          esp_icv_len = 16; // AEAD Mode - ICV is Associated Data
+          crypt_algo_libgcrypt = GCRY_CIPHER_CHACHA20;
+          crypt_mode_libgcrypt = GCRY_CIPHER_MODE_POLY1305;
+
+          /* The counter mode key includes a 4 byte nonce following the key, which is used as the salt */
+          esp_salt_len = 4;
+          esp_encr_key_len -= esp_salt_len;
+
+          if (esp_encr_key_len != gcry_cipher_get_algo_keylen (crypt_algo_libgcrypt))
+          {
+            REPORT_DISSECTOR_BUG("<ESP Preferences> Error in Encryption Algorithm CHACHA20_POLY1305_IIV: Bad Keylen (%u Bits, need %lu)\n",
+                                 esp_encr_key_len * 8, (unsigned long) gcry_cipher_get_algo_keylen (crypt_algo_libgcrypt) * 8);
+            decrypt_ok = FALSE;
+          }
+          else
+            decrypt_using_libgcrypt = TRUE;
+
+          break;
+
         case IPSEC_ENCRYPT_NULL :
         default :
           /* Fix parameters */
@@ -2189,6 +2218,23 @@ dissect_esp(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, void* data _U_)
               err = gcry_cipher_setiv(*cipher_hd, ctr_block, esp_salt_len + esp_iv_len);
             }
           }
+          else if (esp_encr_algo == IPSEC_ENCRYPT_CHACHA20_POLY1305_IIV)
+          {
+            // Implicit IV, see https://www.rfc-editor.org/rfc/rfc8750.html
+            memset(esp_iiv, 0, 12);
+            memcpy(esp_iiv, esp_encr_key + esp_encr_key_len, esp_salt_len);
+            esp_iiv[8] = (sequence_number >> 24) & 0xff;
+            esp_iiv[9] = (sequence_number >> 16) & 0xff;
+            esp_iiv[10] = (sequence_number >> 8) & 0xff;
+            esp_iiv[11] = sequence_number & 0xff;
+            err = gcry_cipher_setiv(*cipher_hd, esp_iiv, 12);
+            // AAD for IIV consists of spi + seq (the first 8 bytes of the esp header)
+            //esp_decr_data_len = esp_encr_data_len - 16; // The last 16 Bytes are the Poly1305 tag
+            //guint8 *aad = wmem_alloc(pinfo->pool, ESP_HEADER_LEN);
+            //tvb_memcpy(tvb, aad, 0, 8);
+            //REPORT_DISSECTOR_BUG("<IPsec/ESP Dissector> AAD: %02x %02x %02x %02x %02x %02x %02x %02x\n", aad[0], aad[1], aad[2], aad[3], aad[4], aad[5], aad[6], aad[7]);
+            //err = gcry_cipher_authenticate(*cipher_hd, aad, ESP_HEADER_LEN);
+          }
           else
           {
             err = gcry_cipher_setiv(*cipher_hd, esp_iv, esp_iv_len);
@@ -2607,6 +2653,7 @@ proto_register_ipsec(void)
   static const value_string esp_proto_type_vals[] = {
     { IPSEC_SA_IPV4, "IPv4" },
     { IPSEC_SA_IPV6, "IPv6" },
+    { IPSEC_SA_ANY, "Any" },
     { 0x00, NULL }
   };
 
